File: <100/011/011.cpp
#include<iostream>
#include<fstream>
using namespace std;

int DR[] = {0, 1, 1,  1};
int DC[] = {1, 0, 1, -1};

int F[25][25], H=20, W=20, answer;
string filename = "011.txt";

int check(int r, int c){
  int Max = 0;
  for(int k = 0; k < 4; ++k){
    int tmp = 1;
    if( r + 3*DR[k] < 0 ) continue;
    if( c + 3*DC[k] < 0 ) continue;
    if( r + 3*DR[k] >= H ) continue;
    if( c + 3*DC[k] >= W ) continue;
    for(int d = 0; d < 4; ++d) tmp *= F[r+d*DR[k]][c+d*DC[k]];
    Max = max(Max, tmp);
  }

  return Max;
}


int main(){
  ifstream in(filename);
  cin.rdbuf(in.rdbuf());

  for(int i = 0; i < H; ++i){
    for(int j = 0; j < W; ++j){
      cin >> F[i][j];
    }
  }

  for(int i = 0; i < H; ++i){
    for(int j = 0; j < W; ++j){
      answer = max(answer, check(i, j));
    }
  }

  cout << answer << endl;
  return 0;
}

File: <100/011/011.py
L = 4
H = 20
W = 20


def solve(F: list[list[int]]) -> int:
    DR = [-1, 0, 1, 1]
    DC = [ 1, 1, 1, 0]
    answer: int = 0

    for r in range(H):
        for c in range(W):
            for k in range(4):
                max_r, max_c = r+(L-1)*DR[k], c+(L-1)*DC[k]
                if max_r < 0:  continue
                if max_c < 0:  continue
                if max_r >= H: continue
                if max_c >= W: continue
                tmp: int = 1
                for d in range(L):
                    r_, c_ = r+d*DR[k], c+d*DC[k]
                    tmp *= F[r_][c_]
                answer = max(answer, tmp)

    return answer


def main() -> None:
    filename = "011.txt"
    F = []
    with open(filename, 'r') as f:
        for row in f:
            if row[-1] == '\n':
                row = row[: -1]
            row_list = [int(c) for c in row.split(' ')]
            F.append(row_list)

    print(solve(F))

    return


if __name__ == '__main__':
    main()

